Fixed cleanup of persistent volumes.

jsafrane opened the pull request:
Kubelet.cleanupOrphanedVolumes() compares list of volumes mounted to a node with list of volumes that are required by pods scheduled on the node ("scheduled volume").

Both lists should contain real volumes, i.e. when a pod uses PersistentVolumeClaim, the list must contain name of the bound volume instead of name of the claim.

Without this patch, the list of scheduled volumes can contain a PersistentVolumeClaim, while the list of mounted volumes contains the volume that is bound to the claim. These lists do not match and the bound volume is then detached from the node.

This happens in a short window of opportunity, when the volume is already attached and before the pod is 'running' (there is additional check to prevent  detaching volume of a running pod). As result, kubelet tries to run a pod which does not have attached all volumes and this fails.

Kubelet cleans up everything and retries to do the same in a moment, but it adds additional delay to pod startup, especially when AWS/GCE/Cinder volume attach/detach is not particularly fast.

markturansky commented:
@saad-ali @thockin this needs to be bumped in priority and review.  

@liggitt @pweil- i think we'll want this for 3.1.1.  It occurred often enough during dev/testing.

@jsafrane nice find.

markturansky commented:
@liggitt @pweil- in conversation with @jsafrane, he thinks this is, actually, not a blocker.  The attached/mounted state will eventually resolve itself, though it might take a few of Kubelet's recon loops to do so.  Bad UX but not a blocker.

markturansky commented:
@jsafrane LGTM.  I think a unit test to expose the bug would help.  You should be able to call `getDesiredVolumes` directly in a test with the correct data payload.  A mock client can return the PV backing a claim.  I use this technique a lot in the various PV controllers.

saad-ali commented:
> @saad-ali @thockin this needs to be bumped in priority and review.

Taking a look

saad-ali commented:
Good find and thanks for the fix @jsafrane!

> I think a unit test to expose the bug would help.

+1

Otherwise LGTM.

saad-ali commented on pkg/kubelet/kubelet.go:
```diff
@@ -1751,13 +1751,41 @@ func (kl *Kubelet) getPullSecretsForPod(pod *api.Pod) ([]api.Secret, error) {
+	claimSource := volume.VolumeSource.PersistentVolumeClaim
+	if claimSource != nil {
+		// resolve real volume behind the claim
+		claim, err := kl.kubeClient.PersistentVolumeClaims(pod.Namespace).Get(claimSource.ClaimName)
```
The `cleanupOrphanedVolumes(...)` method runs as part of the main kubelet sync loop. This introduces a blocking network call in that loop. Not ideal. Worth documenting this in a comment. Also consider filing an issue to move `HandlePodCleanups()` into a separate thread.

jsafrane replied:
`cleanupOrphanedVolumes` was blocking even before, detaching a volume usually takes a second or two.

I added a comment about that and filled https://github.com/kubernetes/kubernetes/issues/19645

saad-ali replied:
Ya, I put the GCE PD detach on a separate thread for exactly that reason. Thanks for opening the issue, this definitely needs cleanup.

jsafrane commented:
> I think a unit test to expose the bug would help.

Looking at current kubelet tests, it is quite lot of work to simulate this error. I'll try, but I'm focusing on https://github.com/kubernetes/kubernetes/issues/19602 now, we're loosing data here.

jsafrane commented:
Squashed and rebased

markturansky commented:
@saad-ali this looked good to you and we both wanted a test, but @jsafrane says it would be quite a bit of work.  Thoughts?

saad-ali commented:
Let's get this merged. We want this for 1.2 (code freeze on Friday). Please file an issue to add unit test later.

jsafrane committed:
Fixed cleanup of persistent volumes.

Kubelet.cleanupOrphanedVolumes() compares list of volumes mounted to a node
with list of volumes that are required by pods scheduled on the node
("scheduled volume").

Both lists should contain real volumes, i.e. when a pod uses
PersistentVolumeClaim, the list must contain name of the bound volume instead
of name of the claim.

jsafrane commented:
There already is an unit test. I rebased the PR and lost LGTM because of that.
